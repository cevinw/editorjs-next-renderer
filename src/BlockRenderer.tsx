import React from 'react';
import AlertBlock from "./BlockParsers/alert";
import ParagraphBlock from "./BlockParsers/paragraph";
import CodeBlock, {CodeLanguage} from "./BlockParsers/code";
import DelimiterBlock from "./BlockParsers/delimiter";
import HeaderBlock from "./BlockParsers/header";
import ImageBlock from "./BlockParsers/image";
import ColumnsBlock from "./BlockParsers/columns";
import QuoteBlock from "./BlockParsers/quote";
import TableBlock from "./BlockParsers/table";
import ListBlock from "./BlockParsers/list";
import ErrorBlock from "./BlockParsers/error";


export interface ArticleRenderProps {
    data: OutputData;
    config?: BlockRendererConfig
}

export interface BlockRendererConfig {
    alert?: { className: string },
    code?: {
        className?: string,
        codeStyle?: { [key: string]: React.CSSProperties }
        languages?: CodeLanguage[],
        showLineNumbers?: boolean,
    },
    table?: { className: string },
    image?: { className: string },
    header?: { className: string },
    paragraph?: { className: string },
    list?: { className: string },
    quote?: { className: string },
    delimiter?: { className: string }
}

export interface OutputBlockData<Data extends object = any> {

    id?: BlockId;
    type: string;
    data: BlockToolData<Data>;
    tunes?: { [name: string]: BlockTuneData };
}

export interface OutputData {

    version?: string;
    time?: number;
    blocks: OutputBlockData[];
}

export type BlockId = string;
export type BlockToolData<T extends object = any> = T;
export type BlockTuneData = any;

const BlockRenderer = ({data, config}: ArticleRenderProps) =>
    data.blocks.map((blockData: OutputBlockData) => parseData(blockData, config));

const parseData = (item: OutputBlockData, config?: BlockRendererConfig): React.JSX.Element => {
    return {
        "code": <CodeBlock key={item.id} item={item} {...config?.code}/>,
        "paragraph": <ParagraphBlock key={item.id} item={item} {...config?.paragraph}/>,
        "header": <HeaderBlock key={item.id} item={item} {...config?.header}/>,
        "alert": <AlertBlock key={item.id} item={item} {...config?.alert}/>,
        "columns": <ColumnsBlock key={item.id} item={item} config={config}/>,
        "quote": <QuoteBlock key={item.id} item={item} {...config?.quote}/>,
        "table": <TableBlock key={item.id} item={item} {...config?.table}/>,
        "list": <ListBlock key={item.id} item={item} {...config?.list}/>,
        "delimiter": <DelimiterBlock key={item.id} item={item} {...config?.delimiter}/>,
        "image": <ImageBlock key={item.id} item={item} {...config?.image}/>
    }[item.type] || <ErrorBlock item={item} key={item.id}/>
}

export default BlockRenderer;