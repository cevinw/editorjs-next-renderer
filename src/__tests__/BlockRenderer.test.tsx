
import {describe, expect, test} from "vitest";
import {render} from "@testing-library/react";
import React from "react";
import BlockRenderer, {BlockRendererConfig} from "../BlockRenderer";
import * as fs from "node:fs";
import {idea} from "react-syntax-highlighter/dist/cjs/styles/hljs";

describe("BlockRenderer tests", () => {
    test("Block renderer snapshot test", async () => {
        const testData = await JSON.parse(fs.readFileSync(__dirname + "/resources/editorjsOutputData.json", "utf8"));
        const {container} = render(<BlockRenderer data={testData.data}/>);
        expect(container).toMatchSnapshot();
    })

    test("Render with custom classes", async () => {
        const testData = await JSON.parse(fs.readFileSync(__dirname + "/resources/editorjsOutputData.json", "utf8"));
        let config: BlockRendererConfig =  {
            image: {
                className: "testClass"
            },
            code: {
                className: "test",
                codeStyle: idea
            }
        }
        const {container} = render(<BlockRenderer data={testData.data} config={config}/>);
        expect(container.querySelector("img")?.className).toBe("testClass");
        expect(container.querySelector(".test")?.className).toBeDefined()
    })

})

