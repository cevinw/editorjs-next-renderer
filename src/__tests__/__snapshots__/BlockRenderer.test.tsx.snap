// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BlockRenderer tests > Block renderer snapshot test 1`] = `
<div>
  <section
    class="grid mt-2"
  >
    <img
      alt=""
      class="rounded justify-self-center dark:dark:opacity-75"
      data-nimg="1"
      decoding="async"
      height="500"
      loading="lazy"
      src="/_next/image?url=http%3A%2F%2Ftest.test.png&w=1080&q=75"
      srcset="/_next/image?url=http%3A%2F%2Ftest.test.png&w=640&q=75 1x, /_next/image?url=http%3A%2F%2Ftest.test.png&w=1080&q=75 2x"
      style="color: transparent;"
      width="500"
    />
  </section>
  <p
    class="mt-1"
  >
    test
  </p>
  <section
    class="text-sm rounded-md overflow-hidden shadow-sm mt-2"
  >
    <pre
      style="display: block; overflow-x: auto; padding: 0.5em; background: rgb(35, 35, 35); color: rgb(230, 225, 220);"
    >
      <code
        style="white-space: pre;"
      >
        <span
          style="color: rgb(109, 156, 190);"
        >
          test
        </span>
      </code>
    </pre>
  </section>
  <ol>
    <li
      class="list-decimal"
    >
      test
    </li>
    <li
      class="list-decimal"
    >
      test2
    </li>
  </ol>
  <p
    class="text-start"
  >
    <span
      class="italic text-lg"
    >
      "
      test
      "
    </span>
    <br />
    <span
      class="opacity-85 text-sm"
    >
      - 
      test
    </span>
  </p>
  <p
    class="mt-1"
  >
    test
  </p>
  <section
    class="grid gap-4 grid-cols-2"
  >
    <div
      class="self-center"
    >
      <p
        class="mt-1"
      >
        test
      </p>
    </div>
    <div
      class="self-center"
    >
      <p
        class="mt-1"
      >
        test
      </p>
    </div>
  </section>
  <table
    class="w-full"
  >
    <tbody>
      <tr>
        <th
          class="first:border-l-0 text-left pl-2 last:border-r-0 border-b-2 border"
        >
          test
        </th>
        <th
          class="first:border-l-0 text-left pl-2 last:border-r-0 border-b-2 border"
        >
          test
        </th>
      </tr>
      <tr>
        <td
          class="first:border-l-0 last:border-r-0 pl-2 border-b border"
        >
          test
        </td>
        <td
          class="first:border-l-0 last:border-r-0 pl-2 border-b border"
        >
          test
        </td>
      </tr>
    </tbody>
  </table>
  <section
    class="p-2 mt-2 px-4 flex bg-opacity-50 shadow-sm rounded-lg bg-gray-300 text-gray-600"
  >
    <div
      class="text-right"
    >
      test
    </div>
  </section>
  <h1>
    test
  </h1>
</div>
`;
